### PR TITLE
Refactor: Move AutoscalingContext to Estimate func

### PR DESCRIPTION
**Pull request by azylinski**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The Estimator Builder should use (wide) AutoscalingContext instead of (specific sub-) params.
That would be consistent with other interfaces, e.g.: all processors, scaleup/scaledown.

#### Does this PR introduce a user-facing change?

```release-note
"NONE"
```

/assign @x13n

### CODE REVIEW

**Comment by vadasambar** [comment]
> The Estimator Builder should use (wide) AutoscalingContext instead of (specific sub-) params.
> That would be consistent with other interfaces, e.g.: all processors, scaleup/scaledown.

Is there any other reason apart from consistency? I don't see why estimator should match processors or scaleup/scaledown when it comes to passing parameters. They are quite different constructs. 

I think passing context to the estimator makes the function definition less prone to changes in the future because you can just use anything from the context (right now you have to pluck the value from the context and pass it to the estimator) but that also means estimator now knows a whole lot more than it should. Do we really need this?

**Comment by azylinski** [comment]
> Is there any other reason apart from consistency? I don't see why estimator should match processors or scaleup/scaledown when it comes to passing parameters. They are quite different constructs.

I'd say that estimator (especially: `estimationAnalyserFunc` and `binpackingLimiter`) are pretty much processors.


> estimator now knows a whole lot more than it should

I agree 100%. IMHO `AutoscalerContext` by itself is more than it should.
I see it as a collection of: options, clients and "global" state, that we pass around. It's not ideal (again, just my opinion), but it's a design decision that was made long time ago.
And again, this is currently the case for all the processors.


> definition less prone to changes in the future because you can just use anything from the context

In essence, that's the direction I'm aiming for.
Current PR is 2nd part of refactor - and there's one more to come, when I'd like to move nodeInfos (or better: TemplateNodesForGroups) to context as well. That's not only a huge code simplification, but also pairs well wth `ClusterStateRegistry` "level".

**Comment by MaciekPytel** [comment]
> > estimator now knows a whole lot more than it should
>
> I agree 100%. IMHO AutoscalerContext by itself is more than it should.

I think this is a decision we already took with introduction of processors and the definition of AutoscalingContext. The idea was that we want to allow cloud providers / advanced users to have broad power to customize CA to fit the specifics of their platform, including arbitrary changes to internal CA logic. 

CA cloudprovider interface expects very little from provider, which makes it relatively easy to add integrations (as demonstrated by the number of integrations we have). The downside is that our interface and our generic logic is essentially a lowest common denominator which makes it very hard to use more advanced cloud provider features (e.g. how long it took for CA integration with MixedInstancesPolicy? How limiting that integration is to this day?). 

The idea of introducing processors was to address this limitation, by giving providers tools to essentially completely change arbitrary parts of CA logic. To this end we've added programming hooks in most critical parts of CA and gave them full access to all the internal state (i.e. AutoscalingContext). The expectation was users forking CA and writing custom processors that can significantly change / override core logic in order to support their specific use-cases that cannot be satisfied purely via implementation of cloudprovider interface. 

To sum up: we designed AutoscalingContext the way it is, specifically to allow people to inject arbitrary and unexpected logic in CA extension points (processors). It can be argued if this is the right approach, but I think pushing AutoscalingContext to any interface that we reasonably expect to be customized by users (e.g. EstimationAnalyserFunc or BinpackingLimiter) is at least consistent with what we've been doing elsewhere in CA codebase.

File: cluster-autoscaler/estimator/binpacking_estimator.go [hunk 1]
```diff
@@ -21,38 +21,34 @@ import (
...
 	"k8s.io/autoscaler/cluster-autoscaler/utils/scheduler"
 	klog "k8s.io/klog/v2"
 	schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
 )
 
 // BinpackingNodeEstimator estimates the number of needed nodes to handle the given amount of pods.
 type BinpackingNodeEstimator struct {
-	predicateChecker       predicatechecker.PredicateChecker
-	clusterSnapshot        clustersnapshot.ClusterSnapshot
+	ctx                    *cactx.AutoscalingContext
```
Comment (MaciekPytel) [inline cluster-autoscaler/estimator/binpacking_estimator.go:32]: Feel free to ignore this comment if it's very hard to implement, but it would be much more consistent with the rest of the codebase to have ctx as a first parameter to Estimate() rather than keep it here. 
Also it seems counterintuitive to persist context. Our implementation happens to re-use same context object rather than recreating it between loops, but I'd rather not have that as an assumption if we can avoid it.

Comment (azylinski) [reply cluster-autoscaler/estimator/binpacking_estimator.go]: Moved to `Estimate()` - your are absolutely right, context should not be stored.

Comment (azylinski) [reply cluster-autoscaler/estimator/binpacking_estimator.go]: And I couldn't agree more about "the love for EstimationContext". If you look closer: 
 1st: this more of a Limiter / Threshold Context 
 2nd: this is the only reason why we have this builder pattern, so the threshold limits could be updated inline

This should be simplified, but as you pointed out, that's out of PR scope.

**Comment by vadasambar** [comment]
> > > estimator now knows a whole lot more than it should
> > 
> > 
> > I agree 100%. IMHO AutoscalerContext by itself is more than it should.
> 
> I think this is a decision we already took with introduction of processors and the definition of AutoscalingContext. The idea was that we want to allow cloud providers / advanced users to have broad power to customize CA to fit the specifics of their platform, including arbitrary changes to internal CA logic.
> 
> CA cloudprovider interface expects very little from provider, which makes it relatively easy to add integrations (as demonstrated by the number of integrations we have). The downside is that our interface and our generic logic is essentially a lowest common denominator which makes it very hard to use more advanced cloud provider features (e.g. how long it took for CA integration with MixedInstancesPolicy? How limiting that integration is to this day?).
> 
> The idea of introducing processors was to address this limitation, by giving providers tools to essentially completely change arbitrary parts of CA logic. To this end we've added programming hooks (processors) in most critical parts of CA and gave them full access to all the internal state (i.e. AutoscalingContext). The expectation was users forking CA and writing custom processors that can significantly change / override core logic in order to support their specific use-cases that cannot be satisfied purely via implementation of cloudprovider interface.
> 
> To sum up: we designed AutoscalingContext the way it is, specifically to allow people to inject arbitrary and unexpected logic in CA extension points (processors). It can be argued if this is the right approach, but I think pushing AutoscalingContext to any interface that we reasonably expect to be customized by users (e.g. EstimationAnalyserFunc or BinpackingLimiter) is at least consistent with what we've been doing elsewhere in CA codebase.

Appreciate the context @MaciekPytel . I will review the PR from the 

> `binpackingLimiter`

On a second look I see why you say estimator and limiter are processors :)

**Comment by vadasambar** [comment]
> > Is there any other reason apart from consistency? I don't see why estimator should match processors or scaleup/scaledown when it comes to passing parameters. They are quite different constructs.
> 
> I'd say that estimator (especially: `estimationAnalyserFunc` and `binpackingLimiter`) are pretty much processors.
> 
> > estimator now knows a whole lot more than it should
> 
> I agree 100%. IMHO `AutoscalerContext` by itself is more than it should. I see it as a collection of: options, clients and "global" state, that we pass around. It's not ideal (again, just my opinion), but it's a design decision that was made long time ago. And again, this is currently the case for all the processors.
> 
> > definition less prone to changes in the future because you can just use anything from the context
> 
> In essence, that's the direction I'm aiming for. Current PR is 2nd part of refactor - and there's one more to come, when I'd like to move nodeInfos (or better: TemplateNodesForGroups) to context as well. That's not only a huge code simplification, but also pairs well wth `ClusterStateRegistry` "level".

On a second look, I see why you say estimator and limiter are processors :)

**Comment by vadasambar** [comment]
> > > estimator now knows a whole lot more than it should
> > 
> > 
> > I agree 100%. IMHO AutoscalerContext by itself is more than it should.
> 
> I think this is a decision we already took with introduction of processors and the definition of AutoscalingContext. The idea was that we want to allow cloud providers / advanced users to have broad power to customize CA to fit the specifics of their platform, including arbitrary changes to internal CA logic.
> 
> CA cloudprovider interface expects very little from provider, which makes it relatively easy to add integrations (as demonstrated by the number of integrations we have). The downside is that our interface and our generic logic is essentially a lowest common denominator which makes it very hard to use more advanced cloud provider features (e.g. how long it took for CA integration with MixedInstancesPolicy? How limiting that integration is to this day?).
> 
> The idea of introducing processors was to address this limitation, by giving providers tools to essentially completely change arbitrary parts of CA logic. To this end we've added programming hooks (processors) in most critical parts of CA and gave them full access to all the internal state (i.e. AutoscalingContext). The expectation was users forking CA and writing custom processors that can significantly change / override core logic in order to support their specific use-cases that cannot be satisfied purely via implementation of cloudprovider interface.
> 
> To sum up: we designed AutoscalingContext the way it is, specifically to allow people to inject arbitrary and unexpected logic in CA extension points (processors). It can be argued if this is the right approach, but I think pushing AutoscalingContext to any interface that we reasonably expect to be customized by users (e.g. EstimationAnalyserFunc or BinpackingLimiter) is at least consistent with what we've been doing elsewhere in CA codebase.

Appreciate the context. Might be a good idea to document this somewhere (feels like tribal knowledge).

**Comment by x13n** [comment]
I just realized I never replied here. Approving, based on the reasoning above. @azylinski can you rebase?

/approve
